--- Day 6: Guard Gallivant ---
The Historians use their fancy device again, this time to whisk you all away to the North Pole prototype suit manufacturing lab... in the year 1518! It turns out that having direct access to history is very convenient for a group of historians.

You still have to be careful of time paradoxes, and so it will be important to avoid anyone from 1518 while The Historians search for the Chief. Unfortunately, a single guard is patrolling this part of the lab.

Maybe you can work out where the guard will go ahead of time so that The Historians can search safely?

You start by making a map (your puzzle input) of the situation. For example:

....#.....
.........#
..........
..#.......
.......#..
..........
.#..^.....
........#.
#.........
......#...
The map shows the current position of the guard with ^ (to indicate the guard is currently facing up from the perspective of the map). Any obstructions - crates, desks, alchemical reactors, etc. - are shown as #.

Lab guards in 1518 follow a very strict patrol protocol which involves repeatedly following these steps:

If there is something directly in front of you, turn right 90 degrees.
Otherwise, take a step forward.
Following the above protocol, the guard moves up several times until she reaches an obstacle (in this case, a pile of failed suit prototypes):

....#.....
....^....#
..........
..#.......
.......#..
..........
.#........
........#.
#.........
......#...
Because there is now an obstacle in front of the guard, she turns right before continuing straight in her new facing direction:

....#.....
........>#
..........
..#.......
.......#..
..........
.#........
........#.
#.........
......#...
Reaching another obstacle (a spool of several very long polymers), she turns right again and continues downward:

....#.....
.........#
..........
..#.......
.......#..
..........
.#......v.
........#.
#.........
......#...
This process continues for a while, but the guard eventually leaves the mapped area (after walking past a tank of universal solvent):

....#.....
.........#
..........
..#.......
.......#..
..........
.#........
........#.
#.........
......#v..
By predicting the guard's route, you can determine which specific positions in the lab will be in the patrol path. Including the guard's starting position, the positions visited by the guard before leaving the area are marked with an X:

....#.....
....XXXXX#
....X...X.
..#.X...X.
..XXXXX#X.
..X.X.X.X.
.#XXXXXXX.
.XXXXXXX#.
#XXXXXXX..
......#X..
In this example, the guard will visit 41 distinct positions on your map.

Predict the path of the guard. How many distinct positions will the guard visit before leaving the mapped area?

In [44]:
with open('input6.txt') as file:
    data = [line.rstrip() for line in file]
grid_width = len(data[0])
grid_height = len(data)
for i, line in enumerate(data):
    try:
        start = (i, line.index('^'))
        break
    except ValueError:
        continue
obstacles = set()
for i, line in enumerate(data):
    for j, char in enumerate(line):
        if char == '#':
            obstacles.add((i, j))

grid_width, grid_height, start, next(iter(obstacles))

(130, 130, (80, 79), (47, 53))

In [45]:
def outside_grid(pos):
    for coord in pos:
        if coord < 0 or coord >= grid_width: # width == height in input
            return True
    return False

In [46]:
from itertools import cycle

directions = cycle(("up", "right", "down", "left"))
facing = next(directions)
visited = set()
visited.add(start)
pos = start
left_grid = False

while True:
    match facing:
        case 'up':
            looking = (pos[0] - 1, pos[1])
        case 'down':
            looking = (pos[0] + 1, pos[1])
        case 'left':
            looking = (pos[0], pos[1] - 1)        
        case 'right':
            looking = (pos[0], pos[1] + 1)
    if outside_grid(looking):
        break
    elif looking in obstacles:
        facing = next(directions)
        continue
    
    pos = looking
    visited.add(pos)
len(visited)


5131

Your puzzle answer was 5131.

The first half of this puzzle is complete! It provides one gold star: *

--- Part Two ---
While The Historians begin working around the guard's patrol route, you borrow their fancy device and step outside the lab. From the safety of a supply closet, you time travel through the last few months and record the nightly status of the lab's guard post on the walls of the closet.

Returning after what seems like only a few seconds to The Historians, they explain that the guard's patrol area is simply too large for them to safely search the lab without getting caught.

Fortunately, they are pretty sure that adding a single new obstruction won't cause a time paradox. They'd like to place the new obstruction in such a way that the guard will get stuck in a loop, making the rest of the lab safe to search.

To have the lowest chance of creating a time paradox, The Historians would like to know all of the possible positions for such an obstruction. The new obstruction can't be placed at the guard's starting position - the guard is there right now and would notice.

In the above example, there are only 6 different positions where a new obstruction would cause the guard to get stuck in a loop. The diagrams of these six situations use O to mark the new obstruction, | to show a position where the guard moves up/down, - to show a position where the guard moves left/right, and + to show a position where the guard moves both up/down and left/right.

Option one, put a printing press next to the guard's starting position:

....#.....
....+---+#
....|...|.
..#.|...|.
....|..#|.
....|...|.
.#.O^---+.
........#.
#.........
......#...
Option two, put a stack of failed suit prototypes in the bottom right quadrant of the mapped area:


....#.....
....+---+#
....|...|.
..#.|...|.
..+-+-+#|.
..|.|.|.|.
.#+-^-+-+.
......O.#.
#.........
......#...
Option three, put a crate of chimney-squeeze prototype fabric next to the standing desk in the bottom right quadrant:

....#.....
....+---+#
....|...|.
..#.|...|.
..+-+-+#|.
..|.|.|.|.
.#+-^-+-+.
.+----+O#.
#+----+...
......#...
Option four, put an alchemical retroencabulator near the bottom left corner:

....#.....
....+---+#
....|...|.
..#.|...|.
..+-+-+#|.
..|.|.|.|.
.#+-^-+-+.
..|...|.#.
#O+---+...
......#...
Option five, put the alchemical retroencabulator a bit to the right instead:

....#.....
....+---+#
....|...|.
..#.|...|.
..+-+-+#|.
..|.|.|.|.
.#+-^-+-+.
....|.|.#.
#..O+-+...
......#...
Option six, put a tank of sovereign glue right next to the tank of universal solvent:

....#.....
....+---+#
....|...|.
..#.|...|.
..+-+-+#|.
..|.|.|.|.
.#+-^-+-+.
.+----++#.
#+----++..
......#O..
It doesn't really matter what you choose to use as an obstacle so long as you and The Historians can put it into position without the guard noticing. The important thing is having enough options that you can find one that minimizes time paradoxes, and in this example, there are 6 different positions you could choose.

You need to get the guard stuck in a loop by adding a single new obstruction. How many different positions could you choose for this obstruction?

In [47]:
# Ok. A few things to note:
# - The new block must be placed on the previous path of the guard (that is be in visited) or it would have no impact whatsoever. so we check all positions in visited minus the starting position
# - when checking, obv we must avoid getting stuck in a loop. How do we recognize we are in a loop? well if we come across the same spot repeatedly. However, visiting the same spot twice in and of itself is not enough, since we can pass it in different directions without getting stuck in a loop.
# - Two ways to address this: We can either 
#       -record the direction we have passed the tile in (prob easiest) or
#       - simply wait until we pass the same spot for the fifth time, then we know for sure (more computeheavy)
# Lets try to implement the first approach

In [ ]:
def creates_loop(obstacles):
    directions = cycle(("up", "right", "down", "left"))
    facing = next(directions)
    pos = start
    visited = set()    
    visited.add((pos, facing))

    while True:
        match facing:
            case 'up':
                looking = (pos[0] - 1, pos[1])
            case 'down':
                looking = (pos[0] + 1, pos[1])
            case 'left':
                looking = (pos[0], pos[1] - 1)        
            case 'right':
                looking = (pos[0], pos[1] + 1)
        if outside_grid(looking):
            return False
        elif looking in obstacles:
            facing = next(directions)
            continue
        else:
            pos = looking
            if (pos, facing) in visited:
                return True
            else:
                visited.add((pos, facing))


In [ ]:
obstacle_candidates = visited.copy()
obstacle_candidates.remove(start) # not technically part of the problem definition, but very much undefined
num_loops = 0
for candidate in obstacle_candidates:
    if creates_loop(obstacles.union({candidate})):
        num_loops += 1
num_loops

1784

That's the right answer! You are one gold star closer to finding the Chief Historian.

You have completed Day 6! You can [Shareon Bluesky Twitter Mastodon] this victory or [Return to Your Advent Calendar].